pca.py

```diff
@@ -1,70 +1,178 @@
+        return torch.mm(X - self.mean_, self.components_.T)
 
-    def fit_transform(self, X):
-        self.fit(X)
-        return self.transform(X)
+    def forward(self, X, check=True):
+        if self.training:
+            return self.fit(X, check)
+        else:
+            if self.N_:
+                return self.transform(X)
+            raise RuntimeError('PCA has not been fitted')
 
-    def inverse_transform(self, Y):
-        assert hasattr(self, "components_"), "PCA must be fit before use."
-        return torch.matmul(Y, self.components_) + self.mean_
 
-if __name__ == "__main__":
+if __name__ == '__main__':
     import numpy as np
-    from sklearn.decomposition import PCA as sklearn_PCA
-    from sklearn import datasets
-    iris = torch.tensor(datasets.load_iris().data)
-    _iris = iris.numpy()
-    devices = ['cpu']
-    if torch.cuda.is_available():
-        devices.append('cuda')
-    for device in devices:
-        iris = iris.to(device)
-        for n_components in (2, 4, None):
-            _pca = sklearn_PCA(n_components=n_components).fit(_iris)
-            _components = torch.tensor(_pca.components_)
-            pca = PCA(n_components=n_components).to(device).fit(iris)
-            components = pca.components_
-            assert torch.allclose(components, _components.to(device))
-            _t = torch.tensor(_pca.transform(_iris))
-            t = pca.transform(iris)
-            assert torch.allclose(t, _t.to(device))
-        __iris = pca.inverse_transform(t)
-        assert torch.allclose(__iris, iris)
-    print("passed!")
+    from sklearn.decomposition import IncrementalPCA as sklearn_IPCA
+    from sklearn.datasets import make_classification
+
+    n_components = 5
+    batch_size = 20
+
+    X, _ = make_classification(n_samples=100, n_features=20, random_state=0)
+    X = torch.tensor(X, dtype=torch.float32)
+
+    n_batches = X.shape[0] // batch_size
+
+    sklearn_ipca = sklearn_IPCA(n_components=n_components)
+
+    ipca = IncrementalPCA(n_components=n_components, n_features=X.shape[-1]).cuda().train()
+
+    for i in range(n_batches):
+        start_index = i * batch_size
+        end_index = start_index + batch_size
+        X_batch = X[start_index:end_index]
+        if i != n_batches - 1:
+            sklearn_ipca.partial_fit(X_batch)
+            ipca.partial_fit(X_batch.cuda())
+
+    ipca.eval()
+    torch.save({'pca': ipca.state_dict()}, 'ipca.pkl')
+    ipca = IncrementalPCA(n_components=n_components, n_features=X.shape[-1]).cuda()
+    ipca.load_state_dict(torch.load('ipca.pkl')['pca'])
+
+    X_reduced_sklearn = sklearn_ipca.transform(X_batch)
+    X_reduced_custom = ipca.transform(X_batch.cuda())
+
+    X_reduced_custom_np = X_reduced_custom.cpu().numpy()
+
+    print("\nSklearn IncrementalPCA transformed data (first 5 samples):\n",
+          X_reduced_sklearn[:5])
+    print("Custom IncrementalPCA transformed data (first 5 samples):\n",
+          X_reduced_custom_np[:5])
```

Comment:
Test by inspection can't be automated, there should be an `allclose` in here.

Reply:
It does appear to pass though so the IncrementalPCA implementation works!

Reply:
Improved, see my latest commit